flang/lib/Lower/OpenMP.cpp

```diff
@@ -87,19 +88,41 @@ genOMP(Fortran::lower::AbstractConverter &absConv,
     auto &firOpBuilder = absConv.getFirOpBuilder();
     auto currentLocation = absConv.getCurrentLocation();
     auto insertPt = firOpBuilder.saveInsertionPoint();
+
+    // Clauses.
+    // FIXME: Add support for other clauses.
+    bool hasNumThreads = false;
+    mlir::Value numThreads;
+
+    const auto &parallelOpClauseList =
+        std::get<Fortran::parser::OmpClauseList>(blockDirective.t);
+    for (const auto &clause : parallelOpClauseList.v) {
+      if (const auto &numThreadsClause =
+              std::get_if<Fortran::parser::OmpClause::NumThreads>(&clause.u)) {
+        // OMPIRBuilder accepts `NUM_THREAD` clause as a `Value`.
+        hasNumThreads = true;
+        numThreads = absConv.genExprValue(
+            *Fortran::semantics::GetExpr(numThreadsClause->v));
+      }
+    }
     llvm::ArrayRef<mlir::Type> argTy;
-    mlir::ValueRange range;
-    llvm::SmallVector<int32_t, 6> operandSegmentSizes(6 /*Size=*/,
-                                                      0 /*Value=*/);
-    // create and insert the operation.
+    Attribute defaultValue, procBindValue;
+    // Create and insert the operation.
+    // Create the Op with empty ranges for clauses that are yet to be lowered.
     auto parallelOp = firOpBuilder.create<mlir::omp::ParallelOp>(
-        currentLocation, argTy, range);
-    parallelOp.setAttr(mlir::omp::ParallelOp::getOperandSegmentSizeAttr(),
-                       firOpBuilder.getI32VectorAttr(operandSegmentSizes));
-    parallelOp.getRegion().push_back(new Block{});
+        currentLocation, argTy, Value(), hasNumThreads ? numThreads : Value(),
```

Comment:
There are different build functions for the parallel Op. One of which takes as input " ::mlir::ValueRange operands". Would it make sense to write a function from clauseList -> operands and use this build function here? See a usage here : https://github.com/llvm/llvm-project/blob/8d2d0e84857cb1f2d01456eb433b5172d3a0772b/mlir/lib/Conversion/OpenMPToLLVM/OpenMPToLLVM.cpp#L29

 ```
void ParallelOp::build(::mlir::OpBuilder &, ::mlir::OperationState &odsState, ::llvm::ArrayRef<::mlir::Type> resultTypes, ::mlir::ValueRange operands, ::llvm::ArrayRef<::mlir::NamedAttribute> attributes ) {
  odsState.addOperands(operands);
  odsState.addAttributes(attributes);
  for (unsigned i = 0; i != 1; ++i)
    (void)odsState.addRegion();
  assert(resultTypes.size() == 0u && "mismatched number of return types");
  odsState.addTypes(resultTypes);
}
```

Reply:
This is the builder I tried previously ending up in a range constructed with NULL type. The reason why it seems legitimate in the scenario you shared is that, here(in shared snippet) you received a well formed Operands as ValueRange whilst in our case you have to construct it on need basis.
And that's the case when a clause was not present(i.e NUM_THREAD) we were ending up creating a ValueRange with NULL type.

Reply:
Why would you push empty Value(). If the operands is not here it shouldn't have value at all. There is a way to do this with a SmallVector of Value that you populate with actual Value when there are present. I'll push a PR for the acc.loop today that use this mechanism.

Reply:
I'd suggest using an `llvm::Optional<mlir::Value>` in your build method for operands that are truly optional. (Although this looks more like stub code to me.)

Reply:
Okay, let me try to clarify. The benefit here using this Empty `Value` and `ValueRange` is that `builder`  and interfacing looks clean and it doesn't hurt rest of the pipeline(It's Okay to pass them).
+ Adding support for further clauses also seems cleaner using this approach. Synthesize the `Value` independently and construct the Op based on this `Value`.
I'll wait for @clementval mentioned approach!

Reply:
Sure this is also a way to do it. If you want to use the basic builder which takes only one `ValueRange` for operands you can have a look at my PR #419. Basically, it just populate a `SmallVector` of `Value` that can be used in the builder directly. You just need to track the number of operands for each segment.

Reply:
Thanks for the comment and sharing your approach @clementval. Sorry but, I didn't find moving to your approach convincing enough for two reasons:
1. Considering the overall(short term) goal. We can't handle/lower all the clauses till the privatization part completes.
2. Your approach want us to track operands for each segment, whilst this one automatically take cares of that since we already created operands carefully and explicitly. Basically by using `vector ` we land back to where we started earlier.

Code will look like:
```
....
operands.push_back();
 auto parallelOp = firOpBuilder.create<mlir::omp::ParallelOp>(currentLocation, argTy, operands);
parallelOp.setAttr(mlir::omp::ParallelOp::getOperandSegmentSizeAttr(),
                firOpBuilder.getI32VectorAttr(
            {/*if_expr_var=*/0, (hasNumThreads ? 1 : 0), /*private_vars=*/0,
             /*first_private_vars=*/0, /*shared_vars=*/0, /*copyin_vars=*/0}));
```
@kiranchandramohan @clementval @kiranktp does above explanation seem justifiable ? Or if you guys still have reservations for above approach. Please confirm, I'll do it :). Tested and ready!

Reply:
> 1. Considering the overall(short term) goal. We can't handle/lower all the clauses till the privatization part completes.
I think this is needed only when you lower from FIR/OpenMP down to LLVM IR but the lowering from the PFT to FIR/OpenMP can be done anytime. 

Fine for me if you want to use the specialized builder. In the end you either track the number of operands in multiple vectors vs. a single vector with some additional variable. My guess is that the single vector approach might be generalized letter to fit several construct lowering but that still to be proven.

Reply:
Can I consider this as approval and straighten/squash up the PR ?
Ofcourse will wait for others to approve, but this seems right time to me for squashing up.